### DISCUSSION TITLE
Is there an example on how to create / use ESM plugins?

### DISCUSSION THREAD

**Discussion by neverendingqs** [serverless/serverless#10701] [Q&A]
Related:
* https://github.com/serverless/serverless/discussions/10629
* https://github.com/serverless/serverless/pull/10657

I gave it a try yesterday, but couldn't get it working. I couldn't use `import` without renaming the plugin file in `.serverless_plugins/` to `.mjs`, but when I did name it with `.mjs`, Serverless couldn't find it.

@frozenbonito, any chance you have an example you can share? I would be interested in both a plugin in `.serverless_plugins/` and one that's a separate `npm` package.

**Comment by frozenbonito** [comment]
@neverendingqs 

`.serverless_plugins/` directory is currently legacy, so ESM plugin is not supported in that.

You have to put the relative path of your plugin into `serverless.yml`.
https://www.serverless.com/framework/docs/guides/plugins/creating-plugins#creating-a-plugin

You can set `"type"` field in `package.json` to `"module"` to enable ESM syntax like `import` or `export`.
In this case, you don't have to change file extension to `.mjs`.
https://nodejs.org/api/packages.html#type

**Comment by frozenbonito** [comment]
Here is an example.

```console
$ tree
.
├── esm-plugin.js
├── package.json
└── serverless.yml

0 directories, 3 files
```

`package.json`:

```json
{
  "type": "module",
  "devDependencies": {
    "serverless": "^3.2.1"
  }
}
```

`esm-plugin.js`:

```javascript
export default class ESMPlugin {
  constructor() {
    this.commands = {
      "esm-print": {
        lifecycleEvents: ["run"],
      },
    };

    this.hooks = {
      "esm-print:run": () => console.log("ESM plugin"),
    };
  }
}
```

`serverless.yml`:

```yaml
service: serverless-esm-plugin

frameworkVersion: "3"

provider:
  name: aws
  runtime: nodejs12.x

plugins:
  - ./esm-plugin
```

Run:

```console
$ npm install
$ npx sls esm-print
ESM plugin
```

For npm, `"main"` field in `package.json` is also required.
Note that ESM plugins require serverless v3.2.0 or later and Node.js v12.22.0 or later.

`package.json`:

```json
{
  "type": "module",
  "main": "esm-plugin.js",
  "peerDependencies": {
    "serverless": "^3.2.0"
  },
  "engines": {
    "node": ">=12.22"
  }
}
```

**Comment by mnapoli** [comment]
@neverendingqs sorry for highjacking this discussion, just wanted to let you know I send you a DM on Twitter (about feature ideas for Serverless Framework). Just wanted to make sure you get a chance to see it.
